velox/type/Variant.cpp

```diff
@@ -187,7 +187,21 @@ std::string variant::toJson(const TypePtr& type) const {
     return '"' + DATE()->toString(value<TypeKind::INTEGER>()) + '"';
   }
 
-  switch (kind_) {
+  // We need to handle and pass the right TypeKind to switch-case based on the
+  // TypePtr received to this function. This special case is applicable to ROW,
+  // ARRAY and MAP type kinds as they call toJson() recursively.
+  TypeKind checkKind;
+  if (type->isRow() == true) {
+    checkKind = TypeKind::ROW;
+  } else if (type->isArray() == true) {
+    checkKind = TypeKind::ARRAY;
+  } else if (type->isMap() == true) {
+    checkKind = TypeKind::MAP;
+  } else {
+    checkKind = kind_;
+  }
+
+  switch (checkKind) {
```

Comment:
`type->kind()` should do here?
add a check `VELOX_CHECK_EQ(type->kind(), kind_, "Type kind mismatch {} != {}",type->kind(), kind_);`

Reply:
I found the issue. We can't use `expected.at(0)->childAt(0)->type()` to get the expectedType. In various cases, we have expected to have 3 different variant rows, i.e, 3 different children and so picking the first one only is not the right solution. We need to change this logic here.

Reply:
It definitely should be should be `expected.at(0)->type()`. You need to include your recursive toJson fix as well.

Reply:
Fixed.